Cache lns records

darcys22 opened the pull request:
Addresses #1077

Creates a cache in the CLI wallet for LNS records.

LNS records are stored in the blockchain with a hashed name and encrypted value, these details can be recreated if the user knows the unhashed name however this requires the user to manually maintain a listing of their LNS names.

This pull request creates a cache in the wallet to store the full LNS records including the unhashed name and unencrypted value. This is updated when the user buys or updates a LNS mapping, and is also updated when a LNS name is searched using `lns_print_name_to_owners`.

This allows for the command `lns_print_owners_to_names` to print the unencrypted value and unhashed name if it has been saved in the cache.

Doy-lee commented on src/wallet/wallet2.h:
```diff
@@ -1044,6 +1044,23 @@ namespace tools
+      std::string owner;
+      std::string backup_owner;
+    };
+    std::vector<lns_detail> lns_records_cache;
```
An unordered_map seems more appropriate, hashing on the base64 string.

darcys22 commented:
Still got a couple minor items to do (printing cache when wallet offline). But changed the vector to an unordered map on the hashed name and, added the code to the RPC commands and implemented the version guard (Bumping to wallet version 30).

darcys22 commented:
So i think the only issue i havent addressed yet is the displaying the cache when the wallet is offline. As the current printouts for the lns commands are at the bottom of the function ill probably be creating a new function to handle the logic for offline daemon because right now all the functions do is return immediately if not connected.

In regards to mapping type it looks like the rpc functions return session names only. Where there has been some work on lokinet types in the cli functions. Is there still work to do on the rpc side? Should i leave it as session only for the time being?

And finally im not sure whats causing that failed test? It looks like a created transaction fails but didnt think my code touched anything that affected a transaction posted to the chain

jagerman commented:
> And finally im not sure whats causing that failed test? It looks like a created transaction fails but didnt think my code touched anything that affected a transaction posted to the chain

Ignore that (or rebase onto current dev), it's nothing to do with this PR.

> In regards to mapping type it looks like the rpc functions return session names only. Where there has been some work on lokinet types in the cli functions. Is there still work to do on the rpc side? Should i leave it as session only for the time being?

AFAIK everything should be there in the RPC in current dev, returned lokinet entries will have a `"type"` value of 2 (or 0 for Session).  Perhaps you need to rebase your PR branch on top of current dev?  Ping me on Telegram if you need help doing that.
